### PR TITLE
fix: add `instantiateMVars` to `replaceLocalDecl`

### PR DESCRIPTION
**Pull request by thorimur**
* [X] Put an X in this bracket to confirm you have read the
  [External Contribution Guidelines](https://github.com/leanprover/lean4/blob/master/doc/contributions.md).

First half of fixing #2711. (Second half is #2728.) Makes `replaceLocalDecl` more robust by inserting an `instantiateMVars`. It should now be less possible to introduce an unknown free variable by `replaceLocalDecl` alone. This affects many `at` tactics.

However, this might not safeguard against all errors—in particular, mvars can still be synthesized to later-occurring fvars after the fact. See issue #2727.

-----

This was benched on mathlib against nightly testing [here](http://speed.lean-fro.org/mathlib4/compare/e7b27246-a3e6-496a-b552-ff4b45c7236e/to/e7b27246-a3e6-496a-b552-ff4b45c7236e?hash1=62bbf6d19813dde7b2487a48393c8831e64dacfd&hash2=32f7c7c8847731c65a954f25f0f82c84c81c1d61). There didn't seem to be significant changes.

To find the relevant mathlib4 branches on speedcenter manually, search "2712" and compare the branch `bench-nightly-testing-for-2712` (baseline) to `lean-pr-testing-2712`. 

Note: it took a while to get mathlib and std synced up to the current lean changes; the mathlib build failures were unrelated to this PR. This PR doesn't require any changes to mathlib or std.

### CODE REVIEW

File: tests/lean/replaceLocalDeclInstantiateMVars.lean [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+and clearing the old one if possible.
+
+It makes sure that the new local declaration is inserted at the soonest point after `fvarId` at
+which `typeNew` is well-formed. It does this by traversing `typeNew` and finding the `FVarId` with
+the maximum index among all `FVarId`s occurring in `typeNew`.
+
+However, if it encounters a metavariable during this traversal, it simply continues traversing,
+even if this metavariable is assigned to an expression which contains an `FVarId` occurring after
+`fvarId`. This can lead to `replaceLocalDecl` inserting a local declaration with a type which
+depends on `FVarId`s which come after it in the local context, thus creating unknown `FvarId`s (the
```
Comment (semorrison) [inline tests/lean/replaceLocalDeclInstantiateMVars.lean:22]:
```suggestion
depends on `FVarId`s which come after it in the local context, thus creating unknown `FVarId`s (the
```

**Comment by thorimur** [comment]
WIP

**Comment by thorimur** [comment]
awaiting-review

File: tests/lean/replaceLocalDeclInstantiateMVars.lean [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+It makes sure that the new local declaration is inserted at the soonest point after `fvarId` at
+which `typeNew` is well-formed. It does this by traversing `typeNew` and finding the `FVarId` with
+the maximum index among all `FVarId`s occurring in `typeNew`.
+
+However, if it encounters a metavariable during this traversal, it simply continues traversing,
+even if this metavariable is assigned to an expression which contains an `FVarId` occurring after
+`fvarId`. This can lead to `replaceLocalDecl` inserting a local declaration with a type which
+depends on `FVarId`s which come after it in the local context, thus creating unknown `FVarId`s (the
+`FVarId`s of the local declarations after the insertion are modified).
```
Comment (semorrison) [inline tests/lean/replaceLocalDeclInstantiateMVars.lean:23]: Can you clarify that this is describing the previous behaviour (which would have caused this test to fail)?

Comment (thorimur) [reply tests/lean/replaceLocalDeclInstantiateMVars.lean]: Okay, done in fbc8de6.

**Pull request merged**